File: scratch/compact-guards-integration-20260916/compact/src/Test/PrimesFFI.cc
#include "purescript.h"
#include <functional>
#include <memory>

namespace {
    struct Node;
    using List = std::shared_ptr<const Node>;
    struct Node { int head; List tail; };

    List cons(int head, List tail) {
        return std::make_shared<Node>(Node{head, tail});
    }

    List range(int start, int end) {
        List result;
        for (int value = end; value >= start; --value) {
            result = cons(value, result);
        }
        return result;
    }

    List reverse(List values) {
        List result;
        while (values) {
            result = cons(values->head, result);
            values = values->tail;
        }
        return result;
    }

    List filter(const std::function<bool(int)>& predicate, List values) {
        List accepted;
        while (values) {
            if (predicate(values->head)) accepted = cons(values->head, accepted);
            values = values->tail;
        }
        return reverse(accepted);
    }

    List sieve(List values) {
        if (!values) return List{};
        const int prime = values->head;
        return cons(prime, sieve(filter(
            [prime](int value) { return value % prime != 0; }, values->tail)));
    }

    int sumList(List values) {
        int sum = 0;
        while (values) {
            sum += values->head;
            values = values->tail;
        }
        return sum;
    }
}

FOREIGN_BEGIN(Test_PrimesFFI)
exports["runPrimesFFI"] = [](const boxed& in) -> boxed {
    const int result = sumList(sieve(range(2, unbox<int>(in))));
    return result;
};
FOREIGN_END

File: scratch/purust-fbip-20260916/benchmark/src/Test/PrimesFFI.java
private static final class IntList {
    final int head;
    final IntList tail;

    IntList(int head, IntList tail) {
        this.head = head;
        this.tail = tail;
    }
}

private static IntList range(int start, int end) {
    IntList result = null;
    for (int current = end; current >= start; current--) {
        result = new IntList(current, result);
    }
    return result;
}

private static IntList reverse(IntList list) {
    IntList result = null;
    for (; list != null; list = list.tail) {
        result = new IntList(list.head, result);
    }
    return result;
}

private static IntList filter(java.util.function.IntPredicate predicate, IntList list) {
    IntList result = null;
    for (; list != null; list = list.tail) {
        if (predicate.test(list.head)) result = new IntList(list.head, result);
    }
    return reverse(result);
}

private static IntList sieve(IntList list) {
    if (list == null) return null;
    int prime = list.head;
    return new IntList(prime, sieve(filter(x -> x % prime != 0, list.tail)));
}

public static final java.util.function.Function<Object, Object> runPrimesFFI = Bench.nativeBenchmark(input -> {
    IntList primes = sieve(range(2, ((Number) input).intValue()));
    int sum = 0;
    for (; primes != null; primes = primes.tail) sum += primes.head;
    return sum;
});
